Add API guide (#620)
scalar-labs/scalardb: docs/api-guide.md

> # Java API Guide

Author: Added Java API Guide.

# Java API Guide

> This guide briefly explains how to use Scalar DB Java API.

Reviewer: I think it's helpful to have indexes like below.
```
* [Administrative operations](link)
* [Transactional operations](link)
```

Reviewer: @brfrn169 Thank you for the update. But, the link seems not working.

Author: @feeblefakie Sorry for that... 😞 I fixed the broken links. Please take a look again when you have time!

## Administrative operations

> This chapter explains how to execute administrative operations in Scalar DB.

Reviewer: ```suggestion
This section explains how to execute administrative operations in Scalar DB.  You can execute administrative operations programmatically as follows, but you can also execute those operations through [the Schema Loader](link).
```

### Get a DistributedTransactionAdmin instance

To execute administrative operations, you first need to get a `DistributedTransactionAdmin` instance.
The `DistributedTransactionAdmin` instance can be obtained from `TransactionFactory` as follows:

```java
TransactionFactory transactionFactory = TransactionFactory.create("<configuration file path>");
DistributedTransactionAdmin admin = transactionFactory.getTransactionAdmin();
```

Please see [Getting Started](getting-started.md) for the details of the configuration file.

### Create a namespace

Before creating tables, namespaces must be created since a table belongs to one namespace.
You can create a namespace as follows:

```java

> // Create a namespace "ns"
> admin.createNamespace("ns");
> 
> // Create a namespace if not exists
> boolean ifNotExists = true;
> admin.createNamespace("ns", ifNotExists);

Reviewer 2: ```suggestion
// Create a namespace "ns". It will throw an exception if the namespace already exists
admin.createNamespace("ns");

// Create a namespace only if it does not already exist
boolean ifNotExists = true;
admin.createNamespace("ns", ifNotExists);
```
What do you think?

| compaction-strategy  | Cassandra compaction strategy, must be `LCS`, `STCS` or `TWCS`                        | `STCS`           |
| replication-factor   | Cassandra replication factor                                                          | 1                |

### Create a table

Next, we will discuss table creation.

You firstly need to create the TaleMetadata as follows:

```java
// Define a table metadata
TableMetadata tableMetadata =
    TableMetadata.newBuilder()
        .addColumn("c1", DataType.INT)
        .addColumn("c2", DataType.TEXT)
        .addColumn("c3", DataType.BIGINT)
        .addColumn("c4", DataType.FLOAT)
        .addColumn("c5", DataType.DOUBLE)
        .addPartitionKey("c1")
        .addClusteringKey("c2", Scan.Ordering.Order.DESC)
        .addClusteringKey("c3", Scan.Ordering.Order.ASC)
        .addSecondaryIndex("c4")
        .build();
```

Here you define columns, a partition key, a clustering key including clustering orders, and secondary indexes of a table.

Please see [Scalar DB design document - Data Model](design.md#Data Model) for the details of the Scalar DB Data Model.

And then, you can create a table as follows: 

```java

> // Create a table "ns.tbl"
> admin.createTable("ns", "tbl", tableMetadata);
> 
> // Create a table if not exists
> boolean ifNotExists = true;
> admin.createTable("ns", "tbl", tableMetadata, ifNotExists);

Reviewer 2: ```suggestion
// Create a table "ns.tbl". It will throw an exception if the table already exists
admin.createTable("ns", "tbl", tableMetadata);

// Create a table only if it does not already exist
boolean ifNotExists = true;
admin.createTable("ns", "tbl", tableMetadata, ifNotExists);
```

### Create a secondary index

you can create a secondary index as follows:

```java
// Create a secondary index on a column "c5" of a table "ns.tbl". It will throw an exception if the secondary index already exists
admin.createIndex("ns", "tbl", "c5");

// Create a secondary index only if it does not already exist
boolean ifNotExists = true;
admin.createIndex("ns", "tbl", "c5", ifNotExists);

// Create a secondary index with options
Map<String, String> options = ...;
admin.createIndex("ns", "tbl", "c5", options);
```

### Truncate a table

you can truncate a table as follows:

```java
// Truncate a table "ns.tbl"
admin.truncateTable("ns", "tbl");
```

### Drop a secondary index

you can drop a secondary index as follows:

```java
// Drop a secondary index on a column "c5" of a table "ns.tbl". It will throw an exception if the secondary index does not exist
admin.dropIndex("ns", "tbl", "c5");

// Drop a secondary index only if it exists
boolean ifExists = true;
admin.dropIndex("ns", "tbl", "c5", ifExists);
```

### Drop a table

you can drop a table as follows:

```java
// Drop a table "ns.tbl". It will throw an exception if the table does not exist
admin.dropTable("ns", "tbl");

> // Drop a table if exists only if it exists

Reviewer 2: ```suggestion
// Drop a table only if it exists
```

you can get a table metadata as follows:

```java
// Get a table metadata of "ns.tbl"
TableMetadata tableMetadata = admin.getTableMetadata("ns", "tbl");
```

### Operations for Coordinator tables

Depending on the transaction manager type, you need to create coordinator tables to execute transactions. 
The following sections describe the operations for the coordinator table.

#### Create Coordinator tables

You can create coordinator table as follows:

```java
// Create coordinator tables
admin.createCoordinatorTables();

// Create coordinator tables if not exist
boolean ifNotExist = true;
admin.createCoordinatorTables(ifNotExist);

// Create coordinator tables with options
Map<String, String> options = ...;
admin.createCoordinatorTables(options);
```

#### Truncate Coordinator tables

You can create truncate table as follows:

```java
// Truncate coordinator tables
admin.truncateCoordinatorTables();
```

#### Drop Coordinator tables

You can drop truncate table as follows:

```java
// Drop coordinator tables
admin.dropCoordinatorTables();

// Drop coordinator tables if exist
boolean ifExist = true;
admin.dropCoordinatorTables(ifExist);
```

## Transactional operations

> This chapter explains how to execute transactional operations in Scalar DB.

Reviewer: ```suggestion
This section explains how to execute transactional operations in Scalar DB.
```

          // Aborting the transaction fails. You can log it here
        }
        return;
      } catch (UnknownTransactionStatusException e) {
        // If you catch `UnknownTransactionException`, you are not sure if the transaction succeeds
        // or not. In such a case, you need to check if the transaction is committed successfully
        // or not and retry it if it failed. How to identify a transaction status is delegated to
        // users
        return;
      }
    }
  }
}
```

The APIs for CRUD operations (`get()`/`scan()`/`put()`/`delete()`/`mutate()`) could throw `CrudException` and `CrudConflictException`.
If you catch `CrudException`, it indicates some failure (e.g., database failure and network error) happens during a transaction, so you should cancel the transaction or retry the transaction after the failure/error is fixed.
If you catch `CrudConflictException`, it indicates conflicts happen during a transaction so that you can retry the transaction, preferably with well-adjusted exponential backoff based on your application and environment.
The sample code retries three times maximum and sleeps 100 milliseconds before retrying the transaction.

> Also, the `commit()` API could throw `CommitException`, `CommitConflictException`, and `UnknownTransactionException`.

Reviewer 2: ```suggestion
Also, the `commit()` API could throw `CommitException`, `CommitConflictException`, and `UnknownTransactionStatusException`.
```

If you catch `CommitException`, like the `CrudException` case, you should cancel the transaction or retry the transaction after the failure/error is fixed.
If you catch `CommitConflictException`, like the `CrudConflictException` case, you can retry the transaction.

> If you catch `UnknownTransactionException`, you are not sure if the transaction succeeds or not.

Reviewer 2: ```suggestion
If you catch `UnknownTransactionStatusException`, you are not sure if the transaction succeeds or not.
```

# Java API Guide

> This guide briefly explains how to use Scalar DB Java API.

Reviewer: Sorry, it might be better to say something like this since showing the bullet points is a bit sudden. 
```suggestion
Scalar DB Java API is mainly composed of Administrative API and Transactional API. This guide briefly explains what kind of APIs exist and how to use them.
```

> * [Administrative operations](#administrative-operations)

Reviewer: I just realized that we should probably name them like `Administrative API` and `Transactional API` since it's an API guide?
```suggestion
* [Administrative API](#administrative-api)
```

# Java API Guide

Scalar DB Java API is mainly composed of Administrative API and Transactional API.
This guide briefly explains what kind of APIs exist and how to use them.

* [Administrative API](#administrative-api)
* [Transactional API](#transactional-api)

## Administrative API

> This section explains how to execute administrative operations in Scalar DB.

Reviewer: ```suggestion
This section explains how to execute administrative operations with Administrative API in Scalar DB.
```

you can get a table metadata as follows:

```java
// Get a table metadata of "ns.tbl"
TableMetadata tableMetadata = admin.getTableMetadata("ns", "tbl");
```

### Operations for Coordinator tables

Depending on the transaction manager type, you need to create coordinator tables to execute transactions. 
The following items describe the operations for the coordinator table.

#### Create Coordinator tables

You can create coordinator table as follows:

```java
// Create coordinator tables
admin.createCoordinatorTables();

// Create coordinator tables if not exist
boolean ifNotExist = true;
admin.createCoordinatorTables(ifNotExist);

// Create coordinator tables with options
Map<String, String> options = ...;
admin.createCoordinatorTables(options);
```

#### Truncate Coordinator tables

You can create truncate table as follows:

```java
// Truncate coordinator tables
admin.truncateCoordinatorTables();
```

#### Drop Coordinator tables

You can drop truncate table as follows:

```java
// Drop coordinator tables
admin.dropCoordinatorTables();

// Drop coordinator tables if exist
boolean ifExist = true;
admin.dropCoordinatorTables(ifExist);
```

## Transactional API

> This section explains how to execute transactional operations in Scalar DB.

Reviewer: ```suggestion
This section explains how to execute transactional operations with Transactional API in Scalar DB.
```
